File: core/src/main/java/com/lomekwi/cave/pipeline/Frame.java
package com.lomekwi.cave.pipeline;

import com.lomekwi.cave.timeline.Track;

public abstract class Frame implements AutoCloseable{
    public final Track track;
    public volatile long timestamp;
    private final Source<?> source;

    public Frame(Track track) {
        this.track = track;
        this.source = null;
    }

    public Frame(Track track, Source<?> source) {
        this.track = track;
        this.source = source;
    }

    public Source<?> getSource() {
        return source;
    }

    public Frame withTime(long timestamp){
        this.timestamp = timestamp;
        return this;
    }
    @Override
    public void close() {}
}

File: core/src/main/scala/com/lomekwi/cave/pipeline/Frame.scala
package com.lomekwi.cave.pipeline

import java.io.Serializable

@SerialVersionUID(1L)
abstract class Frame(final val trackIndex: Int, private val segment: Segment[?]) extends AutoCloseable with Serializable {

  @volatile var timestamp: Long = 0L
  @volatile private var closed: Boolean = false

  def this(trackIndex: Int) = {
    this(trackIndex, null)
  }

  def getSegment: Segment[?] = segment

  def withTime(timestamp: Long): Frame = {
    this.timestamp = timestamp
    this
  }
  override def close(): Unit = {
    closed = true
  }

  def isClosed: Boolean = closed
}
